### ISSUE TITLE
Times move when timezone changes (eg. daylight savings)

### ISSUE THREAD

**Issue by AlbeyAmakiir** [SamAmco/track-and-graph#214]
I have a graph showing the times in the day I generally do a thing. When daylight savings time switched over, the entire graph shifted left by 1 hour. The data collected before the switch all says `(-1)` next to the timestamp, and the data points after midnight have jumped to the previous day. Since I care about what time of day the thing happens regardless of timezone, this means the graph is not really useful or accurate for me any more.

I'd like the option on a graph to compare times relatively based on the local time for each datapoint rather than from the perspective of a single timezone across the entire dataset.

My guess is that timezone data per point is already kept, since my new points don't say `(-1)`, so *hopefully* this is just a display issue, not a data one.

**Comment by SamAmco** [comment]
I can half fix this. So yes the data contains the relevant information (the local date time that the data was tracked and the offset from UTC when it was tracked) .. The X-Axis on the graph currently just takes the start and end point of the time range, subdivides it into 10 and then calculates what times each of those 10 divisions would represent. It currently does not account for daylight savings (which I can fix) but fundamentally there is a disconnect between the X-Axis of the graph and the data points timestamps. This is because you could track data in multiple different time zones and then plot it all on the same graph. The X-Axis can only show one time zone, so it will pick your current one and translate all timestamps relative to that. 

Tldr; I think i can fix this display issue for you.

**Comment by SamAmco** [comment]
Interesting i'll try and take a look at some point

**Comment by AlbeyAmakiir** [comment]
Oh, in case it helps, I'm in the southern hemisphere. DST ended on the 2nd of April here, while in other parts of the world it would have instead began (and on a slightly different date).

**Comment by SamAmco** [comment]
Ok again here I think i have a partial fix for you. 

So fundamentally the data points are not stored with zone information. This means that for example it's impossible to determine exactly which day of the week/month/year each data point was tracked on in the time zone it was tracked in. This means the data points must be converted to the current time zone before they are iterated for the time histogram. 

Actually probably worse than this there is some inevitable rounding done for histograms. For example for a month there are always 30 bins but not every month has 30 days. So yeah these time histograms are a little rough at times. 

With that said for your specific use case I think i have some sort of solution by using: 

`            dp.timestamp.atZoneSimilarLocal(ZoneId.systemDefault())
`
In other words we can use the time at the local time tracked (i.e. the time it said on the device clock when you tracked the data point) but in the current time zone. 

The upshot is that specifically for the 24 hour case the local times are used so these two data points: 

![image](https://user-images.githubusercontent.com/5012146/235322568-189a7189-39fb-4a7f-a91d-ecbd6a31bee1.png)

Show like this: 

![image](https://user-images.githubusercontent.com/5012146/235322589-3f439539-ced6-4b5e-9f4c-5e3c1b78c86a.png)

Rather than this: 

![image](https://user-images.githubusercontent.com/5012146/235322620-b04ae0b9-c596-4469-9a70-9ee29f2b6883.png)

This will affect histograms for other window sizes as well but I suspect it will only affect them in a way most people find preferable. To be honest though it's hard to get your head around all the possibilities.

**Comment by AlbeyAmakiir** [comment]
Time always manages to find new ways to cause problems for programmers... <_<

Thanks for tracking this down. I'll let you know how it goes on my end on release.

**Comment by AlbeyAmakiir** [comment]
Looks like it's working as of 3.3.1! Thanks much! 😃
